rust-symcrypt/src/ecc/ecdsa.rs

```diff
@@ -0,0 +1,231 @@
+//! EcDsa functions. For further documentation please refer to symcrypt.h
+//!
+//! # Example
+//!
+//! ## ECDSA Sign and Verify
+//!
+//! ```rust
+//! use symcrypt::ecc::{EcKey, CurveType, EcKeyUsage};
+//! use hex::*;
+//!
+//! // Set up input hash value
+//! let hashed_message = hex::decode("4d55c99ef6bd54621662c3d110c3cb627c03d6311393b264ab97b90a4b15214a5593ba2510a53d63fb34be251facb697c973e11b665cb7920f1684b0031b4dd370cb927ca7168b0bf8ad285e05e9e31e34bc24024739fdc10b78586f29eff94412034e3b606ed850ec2c1900e8e68151fc4aee5adebb066eb6da4eaa5681378e").unwrap();
+//!
+//! // Generate a new ECDSA key pair
+//! let key = EcKey::generate_key_pair(CurveType::NistP256, EcKeyUsage::EcDsa).unwrap();
+//!
+//! // Sign the hash value
+//! let signature = key.ecdsa_sign(&hashed_message).unwrap();
+//!
+//! // Verify the signature
+//! let result = key.ecdsa_verify(&signature, &hashed_message);
+//!
+//! // Assert the signature is valid
+//! assert!(result.is_ok());
+//!
+//! ```
+//!
+use crate::ecc::{curve_to_num_format, CurveType, EcKey, EcKeyUsage};
+use crate::errors::SymCryptError;
+use std::vec;
+use symcrypt_sys;
+
+impl EcKey {
+    /// `ecdsa_sign()` returns a signature as a `Vec<u8>`, or a [`SymCryptError`] if the operation fails.
+    ///
+    /// `hashed_message` is a `&[u8]` that represents the hash value to sign.
+    ///
+    /// If the key usage is not [`EcKeyUsage::EcDsa`], or [`EcKeyUsage::EcDhAndEcDsa`] the function will
+    /// fail with a [`SymCryptError`] with the value [`SymCryptError::InvalidArgument`].
+    pub fn ecdsa_sign(&self, hash_value: &[u8]) -> Result<Vec<u8>, SymCryptError> {
+        // SymCrypt code AV's with SymCrypt Fatal Error.
+        // Panic'ing is not normal here in Rust so we are handling the error instead and returning SymCryptError::InvalidArgument
+        if self.get_ec_curve_usage() == EcKeyUsage::EcDh {
+            return Err(SymCryptError::InvalidArgument);
+        }
+        // SymCrypt code AV's with SymCrypt Fatal Error for trying to sign with Curve25519 as it is not supported.
+        // Panic'ing is not normal here in Rust so we are handling the error instead and returning SymCryptError::InvalidArgument
+        if self.get_curve_type() == CurveType::Curve25519 {
+            return Err(SymCryptError::InvalidArgument);
+        }
```

Comment:
Nice find--this is arguably a bug in SymCrypt.

Curve25519 cannot be used with ECDSA, so it's expected that signing doesn't work. But I don't think it should fatal. As the comments say, SymCrypt is a "low-level" library so we tend to leave a lot of parameter validation to higher layers for performance reasons. But checking the curve type before signing (or perhaps at the time of key generation/import) would not be costly relative to the cost of the cryptographic operations.

Anyway, best to leave this here for now, but I'll file a bug on this.

Reply:
Thanks, I'll leave as is for now and we can file an issue and fix it if and when symcrypt fixes the bug